### PR TITLE
Article for extracting model coefficients

### PR DESCRIPTION
**Pull request by topepo**

### CODE REVIEW

File: content/learn/models/coefficients/index.Rmarkdown [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+## A simple linear regression
+
+Let's start with a simple linear regression model: 
+
+$$\hat{y} = \hat{\beta}_0 + \hat{\beta}_1x_1 + \ldots + \hat{\beta}_px_p$$ 
+
+The $\beta$ values are the coefficients and the $x_j$ are model predictors (a.k.a. features). 
+
+For the example, we'll use the [Chicago train data]() where we try to predict the ridership at the Clark and Lake station (column name: `ridership`) with the previous ridership data 14 days prior from two weeks earlier at three stations.
```
Comment (EmilHvitfeldt) [inline content/learn/models/coefficients/index.Rmarkdown]: Missing link for `Chicago train data`

File: content/learn/models/coefficients/index.Rmarkdown [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+pkgs <- c("tidymodels", "glmnet")
+library(Matrix)
+library(glmnet)
+```
+
+## Introduction 
+
+There are many types of statistical models. Some of them have coefficients (a.k.a. weights) for each term in the model. Good examples of these models are linear or logistic regression but more complex models (e.g. neural networks, MARS) can also have model coefficients. There is always interest in examining the estimated coefficients. 
+
+In this article describes how to retrieve the estimated coefficients from models when fit using tidymodels.
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
This article describes how to retrieve the estimated coefficients from models when fit using tidymodels. 
```

File: content/learn/models/coefficients/index.Rmarkdown (hunk 1 above)
Comment (hfrick) [reply content/learn/models/coefficients/index.Rmarkdown]: the lag info is in this sentence twice (14 days vs 2 weeks). are we predicting ridership at a specific station or on the whole line? if it's at a station, are we using the station's lagged measurements?

File: content/learn/models/coefficients/index.Rmarkdown [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+```{r bootstraps}
+set.seed(123)
+bt <- bootstraps(Chicago, times = 5)
+```
+
+With resampling, we fit the same model to the different versions of the data set produced by resampling. The tidymodels function [`fit_resamples()`](https://www.tmwr.org/resampling.html#resampling-performance) is the recommended approach for doing so. 
+
+`fit_resamples()` does not automatically save the model objects for each resample since these can be quite large. However, we can pass a function to `fit_resamples()` that can save the model object or any other aspect of the fit. 
+
+The function takes a single argument that represents the fitted [workflow object](https://www.tmwr.org/workflows.html) (even if you don't give `fit_resamples()` a workflow). We can extract the model fit.
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
The function takes a single argument that represents the fitted [workflow object](https://www.tmwr.org/workflows.html) (even if you don't give `fit_resamples()` a workflow).
```
duplication with the next paragraph

File: content/learn/models/coefficients/index.Rmarkdown [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+This function is then passed to `fit_resamples()`:
+
+```{r lm-resampled}
+lm_res <- 
+  lm_spec %>% 
+  fit_resamples(ridership ~ ., resamples = bt, control = tidy_ctrl)
+lm_res
+```
+
+Note that there is a `.extracts` column in the results. This object contains the results of our `get_lm_coefs()` function. The structure of the elements of this column are a little complex. Let's look at the first element (which corresponds to the first resample):
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
Note that there is a `.extracts` column in the results. This object contains the results of our `get_lm_coefs()` function. The structure of the elements of this column is a little complex. Let's look at the first element (which corresponds to the first resample): 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+lm_res$.extracts[[1]]
+```
+
+There is _another_ column in this element called `.extracts` that has the results of the `tidy()` function call: 
+
+```{r lm-extract-again}
+lm_res$.extracts[[1]]$.extracts[[1]]
+```
+
+These nested columns can flattened via the purrr `unnest()` function:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
These nested columns can be flattened via the purrr `unnest()` function: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+These nested columns can flattened via the purrr `unnest()` function: 
+
+```{r lm-extract-almost}
+lm_res %>% 
+  select(id, .extracts) %>% 
+  unnest(.extracts) 
+```
+
+We still have an column of nested tibbles, so we can run the same command again to get the data into a more useful format:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
We still have a column of nested tibbles, so we can run the same command again to get the data into a more useful format: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+  theme(legend.position = "top")
+```
+
+There seems to be a lot of uncertainty in the coefficient for the Austin station data. 
+
+Looking at the code for unnesting the results, the double-nesting structure may seem excessive. However, the extraction functionality is broad and simpler structures would prevent some use cases. 
+
+## More complex: a glmnet model
+
+The glmnet model can fit the same linear regression model structure shown above. It uses regularization (a.k.a penalization) to estimate the model parameters. This has the benefit of shrinking the coefficients towards zero. This has singificant benefits for situations where there are strong correlations between predictors or if some feature selection is required. Both of these cases is true for our data set.
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
The glmnet model can fit the same linear regression model structure shown above. It uses regularization (a.k.a penalization) to estimate the model parameters. This has the benefit of shrinking the coefficients towards zero. This has singificant benefits for situations where there are strong correlations between predictors or if some feature selection is required. Both of these cases are true for our data set. 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+```{r glmnet-spec}
+glmnet_spec <- 
+  linear_reg(penalty = 0.1, mixture = 0.95) %>% 
+  set_engine("glmnet")
+```
+
+has a penalty that is 95% Lasso and 5% weight decay. The total amount of these two penalties is 0.1 (which is fairly high). 
+
+Due to to how penalization works, the model predictors should be on the same scale. The ridership at our three stations are very different so we will center and scale them using a recipe:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
Due to how penalization works, the model predictors should be on the same scale. The ridership at our three stations are very different so we will center and scale them using a recipe: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+
+Due to to how penalization works, the model predictors should be on the same scale. The ridership at our three stations are very different so we will center and scale them using a recipe: 
+
+```{r recipe}
+rec <- 
+  recipe(ridership ~ ., data = Chicago) %>% 
+  step_normalize(all_numeric_predictors())
+```
+
+These two objects are combined into a model workflow and is also fit:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
These two objects are combined into a model workflow to fit:
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+  add_model(glmnet_spec) %>% 
+  add_recipe(rec)
+
+glmnet_fit <- fit(glmnet_wflow, Chicago)
+glmnet_fit
+```
+
+In this output, the term `lambda` is used to represent the penalty. 
+
+Note that the output shows many values of the penalty despite our specification of `penalty = 0.1`. It turns out that this model fits a continuum of penalty values.  Even though our interest is in a value of 0.1, we can get the model coefficient for many associated values of the penalty form the same model object.
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
Note that the output shows many values of the penalty despite our specification of `penalty = 0.1`. It turns out that this model fits a continuum of penalty values.  Even though our interest is in a value of 0.1, we can get the model coefficient for many associated values of the penalty from the same model object. 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+glmnet_fit <- fit(glmnet_wflow, Chicago)
+glmnet_fit
+```
+
+In this output, the term `lambda` is used to represent the penalty. 
+
+Note that the output shows many values of the penalty despite our specification of `penalty = 0.1`. It turns out that this model fits a continuum of penalty values.  Even though our interest is in a value of 0.1, we can get the model coefficient for many associated values of the penalty form the same model object. 
+
+Let's look at two different approaches to obtaining the coefficients. Both will use the `tidy()` method. One will tidy the glmnet object and the other will tidy tidymodels objects. 
+### Using glmnet penalty values
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown:203]:
```suggestion

### Using glmnet penalty values
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+This glmnet fit contains multiple penalty values for each fit and these depends on the data set; changing the data (or the mixture amount) often produces a different set of values. For this data set, there are `r length(extract_fit_engine(glmnet_fit)$lambda)` penalties available. To get the set of penalties produced for this data set, we can extract the engine fit and tidy:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
This glmnet fit contains multiple penalty values for each fit and these depend on the data set; changing the data (or the mixture amount) often produces a different set of values. For this data set, there are `r length(extract_fit_engine(glmnet_fit)$lambda)` penalties available. To get the set of penalties produced for this data set, we can extract the engine fit and tidy: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+  tidy() %>% 
+  rename(penalty = lambda) %>%   # <- for consistent naming
+  filter(term != "(Intercept)")
+```
+
+This works well but, it turns out that our penalty value (0.1) is not in the list produced by the model. The underlying package has functions that will use interpolation to produce coefficients for this value but the `tidy()` method for glmnet objects does not use it. 
+
+### Using specific penalty values
+
+If we run the `tidy()` method on the workflow or parsnip object, a different function used that returns the coefficients for the penalty value that we specified:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
If we run the `tidy()` method on the workflow or parsnip object, a different function is used that returns the coefficients for the penalty value that we specified: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+
+For any another (single) penalty, we can use an additional argument:
+
+```{r glmnet-tidy-parsnip-alt}
+tidy(glmnet_fit, penalty = 5.5620)  # A value shown above
+```
+
+### Tuning a glmnet model
+
+If we know proper values for penalty and mixture, we can use the `fit_resamples()` function as we did before. Otherwise, we can tune those parameters with the tidymodels `tune_*()` functions.
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]: proper is a slightly weird adjective here? I thought it might mean "valid" until I got to the tuning part.

File: content/learn/models/coefficients/index.Rmarkdown (hunk 13 above)
Comment (hfrick) [reply content/learn/models/coefficients/index.Rmarkdown]: Is it worth explaining why the different tidy methods do different things? ie that (/why) the engine fit doesn't know about the specification of the `penalty` arg to the tidymodels model spec? right now one might read this and think "okay there are two tidy methods and one does what I want and the other does not"

File: content/learn/models/coefficients/index.Rmarkdown [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+We'll tune our glmnet model over both parameters with this grid: 
+
+```{r glmnet-grid}
+pen_vals <- 10^seq(-3, 0, length.out = 10)
+grid <- crossing(penalty = pen_vals, mixture = c(0.1, 1.0))
+```
+
+Here is where some glmnet-related complexity comes in: we know that each resample and each value of `mixture` will probably produce a different set of penalty values contained in the model object. _How can we look at the coefficients at the specific penalty values that we are using to tune?_
+
+The approach that we suggest is to use the special `path_values` option for glmnet. The details are described in the [technical documentation about glmnet and tidymodels](https://parsnip.tidymodels.org/reference/glmnet-details.html#arguments) but this parameter will assign the collection penalty values used by each glmnet fit (irregardless of the data or value of mixture).
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
The approach that we suggest is to use the special `path_values` option for glmnet. The details are described in the [technical documentation about glmnet and tidymodels](https://parsnip.tidymodels.org/reference/glmnet-details.html#arguments) but this parameter will assign the collection of penalty values used by each glmnet fit (regardless of the data or value of mixture). 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+glmnet_tune_spec <- 
+  linear_reg(penalty = tune(), mixture = tune()) %>% 
+  set_engine("glmnet", path_values = pen_vals)
+
+glmnet_wflow <- 
+  glmnet_wflow %>% 
+  update_model(glmnet_tune_spec)
+```
+
+Now we will use a similar extraction function. We add additional argument to retain coefficients that are coerced to zero by the lasso penalty:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
Now we will use a similar extraction function. We add an additional argument to retain coefficients that are shrunk to zero by the lasso penalty: 
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+  glmnet_wflow %>% 
+  tune_grid(
+    resamples = bt,
+    grid = grid,
+    control = parsnip_ctrl
+  )
+glmnet_res
+```
+
+As noted before, the elements of the main `.extracts` column has an embedded list column with the results of `get_glmnet_coefs()`:
```
Comment (hfrick) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
As noted before, the elements of the main `.extracts` column have an embedded list column with the results of `get_glmnet_coefs()`:  
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+
+```{r lm-tidy}
+tidy(lm_fit)
+```
+
+We'll use this function in subsequent sections. 
+
+### Resampled or tuned models
+
+tidymodels emphasizes the use of resampling methods to help evaluated and characterize how well the model works. While time series resampling methods are appropriate for these data, we can also use the [bootstrap](https://www.tmwr.org/resampling.html#bootstrap) to resample the data. This is a standard resampling approach when evaluate the uncertainty in statistical estimates.
```
Comment (EmilHvitfeldt) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
tidymodels emphasizes the use of resampling methods to help evaluated and characterize how well the model works. While time series resampling methods are appropriate for these data, we can also use the [bootstrap](https://www.tmwr.org/resampling.html#bootstrap) to resample the data. This is a standard resampling approach when evaluating the uncertainty in statistical estimates.  
```

File: content/learn/models/coefficients/index.Rmarkdown [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+  theme(legend.position = "top")
+```
+
+There seems to be a lot of uncertainty in the coefficient for the Austin station data. 
+
+Looking at the code for unnesting the results, the double-nesting structure may seem excessive. However, the extraction functionality is broad and simpler structures would prevent some use cases. 
+
+## More complex: a glmnet model
+
+The glmnet model can fit the same linear regression model structure shown above. It uses regularization (a.k.a penalization) to estimate the model parameters. This has the benefit of shrinking the coefficients towards zero. This has singificant benefits for situations where there are strong correlations between predictors or if some feature selection is required. Both of these cases are true for our data set.
```
Comment (EmilHvitfeldt) [inline content/learn/models/coefficients/index.Rmarkdown]:
```suggestion
The glmnet model can fit the same linear regression model structure shown above. It uses regularization (a.k.a penalization) to estimate the model parameters. This has the benefit of shrinking the coefficients towards zero. This has significant benefits for situations where there are strong correlations between predictors or if some feature selection is required. Both of these cases are true for our data set. 
```

**Review by juliasilge — APPROVED** [review]
Really nice addition! 🚀 

We moved this to position 5 on [`/learn/models`](https://www.tidymodels.org/learn/#create-robust-models) and bumped PLS down off the front page.
